src/objects/impls/structure_lab.rs

```diff
@@ -1,11 +1,25 @@
+use stdweb::Value;
+
 use crate::{
     constants::{ResourceType, ReturnCode},
     objects::{Creep, StructureLab},
+    traits::TryFrom,
 };
 
 impl StructureLab {
-    pub fn mineral_type(&self) -> ResourceType {
-        js_unwrap!(__resource_type_str_to_num(@{self.as_ref()}.mineralType))
+    pub fn mineral_type(&self) -> Option<ResourceType> {
+        let mineral_v = js! {
+            const mineral = @{self.as_ref()}.mineralType;
+            if (mineral) {
+                return __resource_type_str_to_num(mineral);
+            }
+        };
+        match mineral_v {
+            Value::Number(_) => {
+                Some(ResourceType::try_from(mineral_v).expect("lab resource unknown."))
+            }
+            _ => None,
```

Comment:
Is it worth converting Value::Undefined to None and panic'ing on any other value?

Reply:
Panics are bad. I'd suggest returning a `Result` if you want to announce an unexpected value.

Reply:
I think `__resource_type_str_to_num` ought throw up an error that'll crash the tick on the js side when a value is present in `mineralType` that doesn't map to numbers the library recognizes, so we shouldn't get bad data at this stage generally; it might be clearer from a code perspective to be explicit about the mapping though?